OptimizelySDK.Tests/OdpTests/OdpSegmentManagerTest.cs

```diff
@@ -0,0 +1,213 @@
+            _mockCache.Setup(c => c.Lookup(Capture.In(keyCollector)))
+                .Returns(default(List<string>));
+            _mockApiManager.Setup(a => a.FetchSegments(It.IsAny<string>(), It.IsAny<string>(),
+                    It.IsAny<OdpUserKeyType>(), It.IsAny<string>(), It.IsAny<List<string>>()))
+                .Returns(segmentsToCheck.ToArray());
+            var manager = new OdpSegmentManager(_odpConfig, _mockApiManager.Object,
+                Constants.DEFAULT_MAX_CACHE_SIZE, null, _mockLogger.Object, _mockCache.Object);
+
+            var segments = manager.FetchQualifiedSegments(FS_USER_ID);
+
+            var cacheKey = keyCollector.FirstOrDefault();
+            Assert.AreEqual(expectedCacheKey, cacheKey);
+            _mockCache.Verify(c => c.Reset(), Times.Never);
+            _mockCache.Verify(c => c.Lookup(cacheKey), Times.Once);
+            _mockLogger.Verify(l =>
+                l.Log(LogLevel.DEBUG, "ODP Cache Miss. Making a call to ODP Server."), Times.Once);
+            _mockApiManager.Verify(
+                a => a.FetchSegments(
+                    API_KEY,
+                    API_HOST,
+                    OdpUserKeyType.FS_USER_ID,
+                    FS_USER_ID,
+                    _odpConfig.SegmentsToCheck), Times.Once);
+            _mockCache.Verify(c => c.Save(cacheKey, It.IsAny<List<string>>()), Times.Once);
+            Assert.AreEqual(segmentsToCheck, segments);
+        }
+
+        [Test]
+        public void ShouldFetchSegmentsSuccessOnCacheHit()
+        {
+            var keyCollector = new List<string>();
+            _mockCache.Setup(c => c.Lookup(Capture.In(keyCollector)))
+                .Returns(segmentsToCheck);
+            _mockApiManager.Setup(a => a.FetchSegments(It.IsAny<string>(), It.IsAny<string>(),
+                It.IsAny<OdpUserKeyType>(), It.IsAny<string>(), It.IsAny<List<string>>()));
+            var manager = new OdpSegmentManager(_odpConfig, _mockApiManager.Object,
+                Constants.DEFAULT_MAX_CACHE_SIZE, null, _mockLogger.Object, _mockCache.Object);
+
+            var segments = manager.FetchQualifiedSegments(FS_USER_ID);
+
+            var cacheKey = keyCollector.FirstOrDefault();
+            Assert.AreEqual(expectedCacheKey, cacheKey);
+            _mockCache.Verify(c => c.Reset(), Times.Never);
+            _mockCache.Verify(c => c.Lookup(cacheKey), Times.Once);
+            _mockLogger.Verify(l =>
+                l.Log(LogLevel.DEBUG, "ODP Cache Hit. Returning segments from Cache."), Times.Once);
+            _mockApiManager.Verify(
+                a => a.FetchSegments(It.IsAny<string>(), It.IsAny<string>(),
+                    It.IsAny<OdpUserKeyType>(), It.IsAny<string>(), It.IsAny<List<string>>()),
+                Times.Never);
+            _mockCache.Verify(c => c.Save(expectedCacheKey, It.IsAny<List<string>>()), Times.Never);
+            Assert.AreEqual(segmentsToCheck, segments);
+        }
```

Comment:
Can we have a test case with empty segmentsToCheck? 
- return empty list
- the request not forwarded to the sever.